crates/storage/provider/src/providers/snapshot/writer.rs

```diff
@@ -0,0 +1,227 @@
+    }
+
+    /// Truncates a number of rows from disk. It delets and loads an older snapshot file if block
+    /// goes beyond the start of the current block range.
+    ///
+    /// ATTENTION: It commits in the end.
+    fn truncate(&mut self, segment: SnapshotSegment, mut num: u64) -> ProviderResult<()> {
+        while num > 0 {
+            let len = match segment {
+                SnapshotSegment::Headers => self.writer.user_header().block_len(),
+                SnapshotSegment::Transactions | SnapshotSegment::Receipts => {
+                    self.writer.user_header().tx_len()
+                }
+            };
+
+            if num >= len {
+                // If there's more rows to delete than this snapshot contains, then just
+                // delete the whole file and go to the next snapshot
+                let previous_snap = self.data_path.clone();
+
+                if self.writer.user_header().block_start() != 0 {
+                    let (writer, data_path) = Self::open(
+                        segment,
+                        self.writer.user_header().block_start() - 1,
+                        self.reader.clone(),
+                    )?;
+                    self.writer = writer;
+                    self.data_path = data_path;
+                }
+
+                NippyJar::<SegmentHeader>::load(&previous_snap)?.delete()?;
+            } else {
+                let to_delete = len - num;
+                self.writer.user_header_mut().prune(to_delete);
+                self.writer.prune_rows(to_delete as usize)?;
+            }
+            num -= len;
+        }
+
+        self.commit()
+    }
+
+    /// Appends column to snapshot file.
+    fn append_column<T: Compact>(&mut self, column: T) -> ProviderResult<()> {
+        self.buf.clear();
+        column.to_compact(&mut self.buf);
+
+        self.writer.append_column(Some(Ok(&self.buf)))?;
+        Ok(())
+    }
+
+    /// Appends transaction to snapshot file.
+    pub fn append_transaction(
+        &mut self,
+        block: BlockNumber,
+        tx_num: TxNumber,
+        tx: TransactionSignedNoHash,
+    ) -> ProviderResult<()> {
+        let segment = SnapshotSegment::Transactions;
+        debug_assert!(self.writer.user_header().segment() == segment);
+
+        self.append(block, segment, tx, |segment_header| {
+            let block_range = segment_header.block_start()..=segment_header.block_end();
+            *segment_header = SegmentHeader::new(block_range, tx_num..=tx_num, segment);
+        })
+    }
+
+    /// Prunes `to_delete` number of transactions from snapshots.
+    pub fn prune_transactions(&mut self, to_delete: u64) -> ProviderResult<()> {
+        let segment = SnapshotSegment::Transactions;
+        debug_assert!(self.writer.user_header().segment() == segment);
+
+        self.truncate(segment, to_delete)
+    }
+}
+
+impl<'a> Drop for SnapshotProviderRW<'a> {
+    fn drop(&mut self) {
+        let _ = self.commit();
+    }
+}
```

Comment:
same as a usual DB transaction, shouldn't this also just abort and not commit the changes?